core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PushTopNIntoTableScan.java

```diff
@@ -43,10 +44,10 @@
 {
     private static final Capture<TableScanNode> TABLE_SCAN = newCapture();
 
-    // Currently the rule is applied at the optimization phase where PARTIAL and FINAL TopNNode do not exist.
-    // The rule can be further made to work with PARTIAL and FINAL if needed.
+    // Rule is executed in two planning phases. Initially we try to pushdown SINGLE TopN into
+    // table scan. If that fails, we repeat the exercise for PARTIAL TopN nodes after SINGLE -> PARTIAL/FINAL split.
     private static final Pattern<TopNNode> PATTERN = topN()
-            .matching(node -> node.getStep().equals(Step.SINGLE))
+            .matching(node -> EnumSet.of(Step.SINGLE, Step.PARTIAL).contains(node.getStep()))
```

Comment:
`node.getStep() == SINGLE || node.getStep() == PARTIAL` seems to be more efficient and equally readable (at least extract a constant)